File: controllers/visiteController.js
const { Visite, Visiteur } = require('../models');

const construireFiltrePérimètre = (req) => {
  const user = req.utilisateur;
  const filtre = {};

  if (!user) return filtre;

  if (user.role === 'SUPER_ADMIN') {
    if (req.query.entrepriseId) filtre.entrepriseId = req.query.entrepriseId;
    if (req.query.agentId) filtre.agentId = req.query.agentId;
  } else if (user.role === 'ADMIN') {
    const entId = user.entrepriseId?._id || user.entrepriseId;
    if (entId) filtre.entrepriseId = entId;
    if (req.query.agentId) filtre.agentId = req.query.agentId;
  } else {
    // AGENT: voit uniquement son propre historique de visites enregistrées
    filtre.agentId = user._id;
  }

  return filtre;
};

const enregistrerEntree = async (req, res) => {
  try {
    const { visiteurId, personneVisitee, service, motif } = req.body;
    const user = req.utilisateur;

    const visiteur = await Visiteur.findById(visiteurId);
    if (!visiteur) return res.status(404).json({ success: false, message: 'Visiteur introuvable.' });

    const visiteEnCours = await Visite.findOne({ visiteurId, statut: 'EN_COURS' });
    if (visiteEnCours) {
      return res.status(409).json({ success: false, message: "Ce visiteur est déjà à l'intérieur.", visiteEnCours });
    }

    const agentId = user?._id || null;
    const entrepriseId = user?.entrepriseId?._id || user?.entrepriseId || null;

    const visite = await Visite.create({
      visiteurId,
      agentId,
      entrepriseId,
      personneVisitee,
      service,
      motif,
      heureEntree: new Date(),
      statut: 'EN_COURS',
    });

    const completeVisite = await Visite.findById(visite._id)
      .populate('visiteurId')
      .populate('agentId', 'nom prenom email')
      .populate('entrepriseId', 'nom code');

    const io = req.app.get('io');
    if (io) {
      io.emit('visite:entree', completeVisite);
    }

    res.status(201).json({
      success: true,
      message: `Entrée enregistrée à ${new Date().toLocaleTimeString('fr-SN')}`,
      visite: completeVisite,
    });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

const enregistrerSortie = async (req, res) => {
  try {
    const visite = await Visite.findById(req.params.id)
      .populate('visiteurId')
      .populate('agentId', 'nom prenom email')
      .populate('entrepriseId', 'nom code');

    if (!visite) return res.status(404).json({ success: false, message: 'Visite introuvable.' });
    if (visite.statut === 'TERMINE') return res.status(400).json({ success: false, message: 'Visite déjà terminée.' });

    visite.heureSortie = new Date();
    visite.statut = 'TERMINE';
    await visite.save();

    const io = req.app.get('io');
    if (io) {
      io.emit('visite:sortie', visite);
    }

    const dureeMinutes = Math.round((new Date() - new Date(visite.heureEntree)) / 60000);
    res.json({ success: true, message: `Sortie enregistrée. Durée : ${dureeMinutes} min.`, visite });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

const listerVisites = async (req, res) => {
  try {
    const { statut, date } = req.query;
    const page  = parseInt(req.query.page)  || 1;
    const limit = parseInt(req.query.limit) || 20;

    const filtre = construireFiltrePérimètre(req);

    if (statut) filtre.statut = statut;
    if (date) {
      const debut = new Date(date);
      const fin   = new Date(date);
      fin.setHours(23, 59, 59, 999);
      filtre.heureEntree = { $gte: debut, $lte: fin };
    }

    const [total, visites] = await Promise.all([
      Visite.countDocuments(filtre),
      Visite.find(filtre)
        .populate('visiteurId')
        .populate('agentId', 'nom prenom email')
        .populate('entrepriseId', 'nom code')
        .sort({ heureEntree: -1 })
        .skip((page - 1) * limit)
        .limit(limit),
    ]);

    res.json({ success: true, total, page, pages: Math.ceil(total / limit), visites });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

const visitesEnCours = async (req, res) => {
  try {
    const filtre = construireFiltrePérimètre(req);
    filtre.statut = 'EN_COURS';

    const visites = await Visite.find(filtre)
      .populate('visiteurId')
      .populate('agentId', 'nom prenom email')
      .populate('entrepriseId', 'nom code')
      .sort({ heureEntree: -1 });

    res.json({ success: true, total: visites.length, visites });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

const supprimerVisite = async (req, res) => {
  try {
    const user = req.utilisateur;
    const visite = await Visite.findById(req.params.id);
    if (!visite) {
      return res.status(404).json({ success: false, message: 'Visite introuvable.' });
    }

    if (user.role === 'AGENT' && String(visite.agentId) !== String(user._id)) {
      return res.status(403).json({ success: false, message: 'Un agent ne peut supprimer que ses propres visites.' });
    }

    if (user.role === 'ADMIN') {
      const entId = user.entrepriseId?._id || user.entrepriseId;
      if (String(visite.entrepriseId) !== String(entId)) {
        return res.status(403).json({ success: false, message: 'Vous ne pouvez supprimer que les visites de votre entreprise.' });
      }
    }

    await Visite.findByIdAndDelete(req.params.id);

    const io = req.app.get('io');
    if (io) {
      io.emit('visite:supprimee', { id: req.params.id });
    }

    res.json({ success: true, message: 'Visite supprimée avec succès.' });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

module.exports = {
  enregistrerEntree,
  enregistrerSortie,
  listerVisites,
  visitesEnCours,
  supprimerVisite,
};

File: controllers/visite_controller.py
from typing import Optional
from datetime import datetime
from fastapi import HTTPException
from pydantic import BaseModel
from bson import ObjectId
from models.visite import Visite
from models.visiteur import Visiteur

class EntreeBody(BaseModel):
    visiteur_id: str
    personne_visitee: str
    service: str
    motif: Optional[str] = None

def _visite_dict(v: Visite) -> dict:
    d = v.dict()
    d["id"] = str(v.id)
    d["visiteur_id"] = str(v.visiteur_id)
    d.pop("_id", None)
    return d

async def enregistrer_entree(body: EntreeBody):
    try:
        oid = ObjectId(body.visiteur_id)
    except Exception:
        raise HTTPException(status_code=400, detail="ID visiteur invalide.")
    visiteur = await Visiteur.get(oid)
    if not visiteur:
        raise HTTPException(status_code=404, detail="Visiteur introuvable.")
    visite_en_cours = await Visite.find_one(Visite.visiteur_id == oid, Visite.statut == "EN_COURS")
    if visite_en_cours:
        raise HTTPException(status_code=409, detail="Ce visiteur est déjà à l'intérieur.")
    visite = Visite(
        visiteur_id=oid,
        personne_visitee=body.personne_visitee,
        service=body.service,
        motif=body.motif,
        heure_entree=datetime.utcnow(),
        statut="EN_COURS",
    )
    await visite.insert()
    d = _visite_dict(visite)
    d["visiteur"] = visiteur.dict()
    return {"success": True, "message": "Entrée enregistrée.", "visite": d}

async def enregistrer_sortie(visite_id: str):
    try:
        oid = ObjectId(visite_id)
    except Exception:
        raise HTTPException(status_code=400, detail="ID invalide.")
    visite = await Visite.get(oid)
    if not visite:
        raise HTTPException(status_code=404, detail="Visite introuvable.")
    if visite.statut == "TERMINE":
        raise HTTPException(status_code=400, detail="Visite déjà terminée.")
    visite.heure_sortie = datetime.utcnow()
    visite.statut = "TERMINE"
    visite.updated_at = datetime.utcnow()
    await visite.save()
    duree_minutes = int((visite.heure_sortie - visite.heure_entree).total_seconds() / 60)
    visiteur = await Visiteur.get(visite.visiteur_id)
    d = _visite_dict(visite)
    if visiteur:
        d["visiteur"] = visiteur.dict()
    return {"success": True, "message": f"Sortie enregistrée. Durée : {duree_minutes} min.", "visite": d}

async def lister_visites(statut: Optional[str] = None, date: Optional[str] = None, page: int = 1, limit: int = 20):
    try:
        # Construction du filtre
        filter_query = {}
        if statut:
            filter_query["statut"] = statut
        if date:
            try:
                debut = datetime.fromisoformat(date + "T00:00:00")
                fin = datetime.fromisoformat(date + "T23:59:59")
                filter_query["heure_entree"] = {"$gte": debut, "$lte": fin}
            except ValueError:
                raise HTTPException(status_code=400, detail="Format de date invalide. Utilisez YYYY-MM-DD.")
        
        # Requête de base
        query = Visite.find(filter_query)
        total = await query.count()
        skip = (page - 1) * limit
        visites = await query.sort(-Visite.heure_entree).skip(skip).limit(limit).to_list()
        
        # Chargement des visiteurs associés
        results = []
        for v in visites:
            d = _visite_dict(v)
            visiteur = await Visiteur.get(v.visiteur_id)
            if visiteur:
                d["visiteur"] = visiteur.dict()
            results.append(d)
        
        return {
            "success": True,
            "total": total,
            "page": page,
            "pages": (total + limit - 1) // limit,
            "visites": results,
        }
    except Exception as e:
        print(f"Erreur dans lister_visites: {e}")
        raise HTTPException(status_code=500, detail=f"Erreur interne: {str(e)}")

async def visites_en_cours():
    try:
        visites = await Visite.find(Visite.statut == "EN_COURS").sort(-Visite.heure_entree).to_list()
        results = []
        for v in visites:
            d = _visite_dict(v)
            visiteur = await Visiteur.get(v.visiteur_id)
            if visiteur:
                d["visiteur"] = visiteur.dict()
            results.append(d)
        return {"success": True, "total": len(results), "visites": results}
    except Exception as e:
        print(f"Erreur dans visites_en_cours: {e}")
        raise HTTPException(status_code=500, detail=f"Erreur interne: {str(e)}")